design-proposals/klip-48-schemas.md

```diff
@@ -0,0 +1,115 @@
+
+## Public APIs
+
+ksqlDB schemas will be primarily leveraged via SQL syntax, with some optional configuration. Each of the new syntax constructs associated with ksqlDB schemas are proposed as follows:
+
+### Creating schemas
+
+```sql
+CREATE SCHEMA my_schema;
+```
+
+This statement will create a new, empty schema. Once a schema is created, it may be used to qualify any newly created objects:
+
+```sql
+CREATE TABLE my_schema.my_table …
+CREATE STREAM my_schema.my_stream …
+CREATE TYPE my_schema.my_type …
+CREATE CONNECTOR my_schema.my_connector …
+
+```
+
+**Any object that can be created and named via `CREATE` should allow an optional schema qualifier.**
+
+### Dropping schemas
+
+```sql
+DROP SCHEMA my_schema;
+```
+
+If `my_schema` is empty, the above `DROP` statement should succeed, at which point `my_schema` should no longer exist. However, if `my_schema` contains any objects, this statement should fail with an error indicating that schemas that contain objects may not be dropped.
+
+If a user would like to drop a schema that contains any objects, they may use the optional `CASCADE` clause:
+
+```sql
+DROP SCHEMA my_schema CASCADE;
+```
+
+This statement should recursively drop all objects contained within `my_schema`, and then drop the schema itself.
+
+### Altering an object’s schema (future)
+
+While it would be worth allowing users to change an object’s schema via `ALTER` syntax, `ALTER` is currently unsupported for connectors and types. As a result, we do not need to support object schema modification in the initial implementation of ksqlDB schema support.
+
+Should we decide to support object schema modification in the future, here are some examples of how it should work:
+
+```sql
+ALTER TABLE my_table SET SCHEMA my_schema;
+ALTER STREAM my_stream SET SCHEMA my_schema;
+ALTER TYPE my_type SET SCHEMA my_schema;
+ALTER CONNECTOR my_connector SET SCHEMA my_schema;
+```
+
+### Configuration
+
+As a matter of convenience and brevity, we should provide a means for users to specify a default schema. If a default schema is given, any references to an object that don’t explicitly include a schema qualifier should be interpreted as targeting the default schema:
+
+```sql
+SET 'ksql.default.schema' = 'my_schema';
+
+-- Targets my_schema.my_table:
+SELECT my_table …
+
+-- Targets schema1.my_stream:
+SELECT schema1.my_stream;
+
+SET 'ksql.default.schema' = 'another_schema';
+
+-- Targets another_schema.my_table:
+SELECT my_table ...
+
+```
+
+## Documentation updates
+
+* New documentation will be required to generally describe ksqlDB schemas and their associated syntax.
+* Any existing documentation pertaining to naming objects will need to be updated to account for the option of qualifying object names with a schema.
+
+## Compatibility implications
+
+* Once users upgrade from a pre-schemas version of ksqlDB to a version that supports schemas, how will ksqlDB interpret the older version's objects' schemas? 
```

Comment:
Cf my comment above: it might be best to have a default schema and migrate existing objects into this default schema when upgrading).